spec/text-editor-presenter-spec.coffee

```diff
@@ -2227,6 +2099,407 @@ describe "TextEditorPresenter", ->
+
+            expectStateUpdate presenter, -> editor.setMini(true)
+            expect(gutterDescriptionWithName(presenter, 'line-number')).toBeUndefined()
+
+            expectStateUpdate presenter, -> editor.setMini(false)
+            expect(gutterDescriptionWithName(presenter, 'line-number').visible).toBe true
+
+            expectStateUpdate presenter, -> editor.setLineNumberGutterVisible(false)
+            expect(gutterDescriptionWithName(presenter, 'line-number').visible).toBe false
+
+            expectStateUpdate presenter, -> editor.setLineNumberGutterVisible(true)
+            expect(gutterDescriptionWithName(presenter, 'line-number').visible).toBe true
+
+            expectStateUpdate presenter, -> atom.config.set('editor.showLineNumbers', false)
+            expect(gutterDescriptionWithName(presenter, 'line-number').visible).toBe false
+
+          it "gets updated when the editor's grammar changes", ->
+            presenter = buildPresenter()
+
+            atom.config.set('editor.showLineNumbers', false, scopeSelector: '.source.js')
+            expect(gutterDescriptionWithName(presenter, 'line-number').visible).toBe true
+            stateUpdated = false
+            presenter.onDidUpdateState -> stateUpdated = true
+
+            waitsForPromise -> atom.packages.activatePackage('language-javascript')
+
+            runs ->
+              expect(stateUpdated).toBe true
+              expect(gutterDescriptionWithName(presenter, 'line-number').visible).toBe false
+
+        it "updates when gutters are added to the editor model, and keeps the gutters sorted by priority", ->
+          presenter = buildPresenter()
+          gutter1 = editor.addGutter({name: 'test-gutter-1', priority: -100, visible: true})
+          gutter2 = editor.addGutter({name: 'test-gutter-2', priority: 100, visible: false})
+          expectedState = [
+            {gutter: gutter1, visible: true},
+            {gutter: editor.gutterWithName('line-number'), visible: true},
+            {gutter: gutter2, visible: false},
+          ]
+          expect(presenter.getState().gutters.sortedDescriptions).toEqual expectedState
+
+        it "updates when the visibility of a gutter changes", ->
+          presenter = buildPresenter()
+          gutter = editor.addGutter({name: 'test-gutter', visible: true})
+          expect(gutterDescriptionWithName(presenter, 'test-gutter').visible).toBe true
+          gutter.hide()
+          expect(gutterDescriptionWithName(presenter, 'test-gutter').visible).toBe false
+
+        it "updates when a gutter is removed", ->
+          presenter = buildPresenter()
+          gutter = editor.addGutter({name: 'test-gutter', visible: true})
+          expect(gutterDescriptionWithName(presenter, 'test-gutter').visible).toBe true
+          gutter.destroy()
+          expect(gutterDescriptionWithName(presenter, 'test-gutter')).toBeUndefined()
+
+      describe ".customDecorations", ->
```

Comment:
Ah, I see. Can you point me to where you take advantage of this ability to key into the structure? I was assuming each gutter component would deal with its own decorations separately.